engine/src/main/java/org/pentaho/di/cluster/SlaveServer.java

```diff
@@ -0,0 +1,1403 @@
+    } finally {
+      lock.readLock().unlock();
+    }
+    CredentialsProvider provider = new BasicCredentialsProvider();
+    UsernamePasswordCredentials credentials = new UsernamePasswordCredentials( userName, password );
+    if ( !Utils.isEmpty( proxyHost ) && host.equals( "localhost" ) ) {
+      host = "127.0.0.1";
+    }
+    provider.setCredentials( new AuthScope( host, port ), credentials );
+    context.setCredentialsProvider( provider );
+    // Generate BASIC scheme object and add it to the local auth cache
+    HttpHost target = new HttpHost( host, port, isSslMode() ? HTTPS : HTTP );
+    AuthCache authCache = new BasicAuthCache();
+    BasicScheme basicAuth = new BasicScheme();
+    authCache.put( target, basicAuth );
+    context.setAuthCache( authCache );
+  }
+
+  private void addProxy( HttpClientContext context ) {
+    String proxyHost;
+    String proxyPort;
+    String nonProxyHosts;
+
+    String hostName;
+
+    lock.readLock().lock();
+    try {
+      proxyHost = environmentSubstitute( this.proxyHostname );
+      proxyPort = environmentSubstitute( this.proxyPort );
+      nonProxyHosts = environmentSubstitute( this.nonProxyHosts );
+
+      hostName = environmentSubstitute( this.hostname );
+    } finally {
+      lock.readLock().unlock();
+    }
+    if ( Utils.isEmpty( proxyHost ) || Utils.isEmpty( proxyPort ) ) {
+      return;
+    }
+    // skip applying proxy if non-proxy host matches
+    if ( !Utils.isEmpty( nonProxyHosts ) && hostName.matches( nonProxyHosts ) ) {
+      return;
+    }
+    HttpHost httpHost = new HttpHost( proxyHost, Integer.valueOf( proxyPort ) );
+
+    RequestConfig requestConfig = RequestConfig.custom()
+      .setProxy( httpHost )
+      .build();
+
+    context.setRequestConfig( requestConfig );
+  }
+
+  /**
+   *
+   * @return HttpClientContext with authorization credentials
+   */
+  protected HttpClientContext getAuthContext() {
+    HttpClientContext context = HttpClientContext.create();
+    addCredentials( context );
+    addProxy( context );
+    return context;
+  }
+
+  /**
+   * @return the master
+   */
+  public boolean isMaster() {
+    lock.readLock().lock();
+    try {
+      return master;
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * @param master the master to set
+   */
+  public void setMaster( boolean master ) {
+    lock.writeLock().lock();
+    this.master = master;
+    lock.writeLock().unlock();
+
+  }
+
+  public String execService( String service, boolean retry ) throws Exception {
+    int tries = 0;
+    int maxRetries = 0;
+    if ( retry ) {
+      maxRetries = KETTLE_CARTE_RETRIES;
+    }
+    while ( true ) {
+      try {
+        return execService( service );
+      } catch ( Exception e ) {
+        if ( tries >= maxRetries ) {
+          throw e;
+        } else {
+          try {
+            Thread.sleep( getDelay( tries ) );
+          } catch ( InterruptedException e2 ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)